### ISSUE TITLE
Bot for ensuring "In Progress" is recent? && Assessing if Kanban is the right methodology for us 

### ISSUE THREAD

**Issue by jbenet** [ipfs/team-mgmt#273]
- **(Background)** Though there may be some flexibility in different projects' definitions, the `In Progress` state is supposed to be accurate about what is actively being worked on in a given moment in time -- not over a large span of time. The point of `In Progress` comes from Kanban, and it is there to give people a sense of what the team is working on _right now_, which helps make collaboration on tasks (and dependencies/potential blockers) smoother. This means that people should seldom (if ever) have two cards on `In Progress`. 
  - I think we **should** use this definition, and we should **commit** to keep it updated this way.
- **(Problem)** Currently, the `In Progress` state is associated with _a lot_ of issues. This is basically just stale. Clearly people are not working on 5+ different things simultaneously. 
  - It is more tedious and error prone to remove the `In Progress` label. We could definitely argue for discipline, but i imagine people are more likely to fail at removing the label, and more likely to succeed at it (because that happens _at the beginning_, when starting to work on something).
- **(Potential solution)** Perhaps we could use a bot to remove the `In Progress` label automatically, after a few hours (say, 3hr, or 6hr, or 12hr), just to make sure staleness is not so much of an issue for everyone else.

What do we think?

**Comment by jbenet** [comment]
---

The algorithm for the bot should be roughly

```py
repos = [ list of repos to check ]
stale = 3 hours # or whatever
for repo in repos:
  in_progress = github.GetIssues(repo=repo, label="In Progress")
  for issue in in_progress:
    if issue.last_modified > stale:
      issue.RemoveLabel("In Progress")
      issue.AddLabel("Ready")

# notes: 
# - maybe "last modified" can be a part of the GetIssues query.
# - of course, the updates can be batched better.
```

**Comment by jbenet** [comment]
By the way-- lots of the project management methodology literature i've read stress heavily the importance of ensuring people are truly only working on one thing at any one time. It helps people focus on that thing for the moment, and let go of others. It helps make sure people truly finish working on one thing (or stop and put it back) before trying to work on another. this is the sort of Project Management discipline that may come off as annoying at first, but can have a transformative effect on our individual and team focus and rate of progress.

**Comment by flyingzumwalt** [comment]
In my experience, I have often had issues that I worked on for a day or two.

Tweak to @jbenet's suggestion: sometimes you want to flag an issue as "I'm going to work on this. Leave it for me to pick up later." In those cases, you should assign the issue to yourself but not mark it "in progress"

**Comment by whyrusleeping** [comment]
@jbenet I would love to start having a management bot. I have a fair number of ideas i'd like to implement with that. This is LGTM, though maybe 3 hours is a bit aggressive

**Comment by flyingzumwalt** [comment]
@jbenet I'm not familiar with the PM approaches that involve forcing people to "truly finish working on one thing (or stop and put it back) before trying to work on another". Please share links to reference info. It sounds like more of a personal productivity habit rather than a PM methodology. Building it into the PM process sets of my alarms about micro-managing.

**Comment by Kubuxu** [comment]
I think it is quite and interesting and nice idea. I also agree with @whyrusleeping that 3h is a bit aggressive, I would be for something like 36h during the week and 62h over the weekend (so "in progress" isn't almost empty on Monday).

**Comment by flyingzumwalt** [comment]
👍 to having a bot
👍 to the time windows @Kubuxu suggested

**Comment by jbenet** [comment]
- 36hrs is WAY too much. this defeats the purpose. 12hr may be ok.
- what is desired here is to be able to look at -- with the board -- at what the team is working on actively. not "today", but "this hour". does this help clarify the intent of this issue + time restriction? im ok starting with 36hr, but the point is to push to higher throughput and only labeling something `In Progress` if you're working on it _right now_, ("this minute" not "today"), and to guide people to _move something back to "Ready" when they're not actively working on it, even if they're "background thinking" about it.
- The point of all this is to make what is being worked on _fully explicit and focused_, and let everything else take a back seat and be pushed out of mind. This allows focus and flow.

This is some reading i found quickly now --. it's not the best at all, but in the interest of time, i link it. i will try to look for the better stuff later. I've read these and they get at some of the points, though it's mostly centered on a corollary / related thing: WIP limits. WIP limits = limit the number of things you can "be working on" at any one time. (usually 1 per person). Again, the emphasis on Kanban is to remove multitasking and to guide towards high task throughput.

WIP limits in Kanban
- http://kanbantool.com/kanban-wip-limits
- https://kanbanize.com/blog/kanban-101-applying-wip-limits/
  - which links to https://hbr.org/2013/03/the-multitasking-paradox
- http://pm.stackexchange.com/questions/11301/what-is-the-process-for-deciding-wip-limits-for-each-column-in-a-teams-kanban-b
- http://www.everydaykanban.com/2013/05/18/true-cost-of-no-wip-limits/
- https://www.equinox.co.nz/blog/kanban-boards-done-columns-and-work-in-progress-limits
- http://itsadeliverything.com/littles-law-the-basis-of-lean-and-kanban
- https://www.smartsheet.com/visualize-your-do-list-using-kanban-boards-optimize-workflow
- http://stackoverflow.com/questions/9392820/what-should-i-do-with-a-kanban-item-that-fails-test

Alternatively, these advocate **NOT** moving items back, in favor of a much more conscientious assignment phase where people are prevented from taking on too much. (i dont think this conscientious assignment will work very well at first with us, because right now people take things for themselves -- it does not go through another person who can prevent the self-assignment -- and people are already taking on many things at once. we need to level up to that so people do it just by themselves and dont need push-back):
- http://zsoltfabok.com/blog/2010/09/never-move-items-back/
- http://brodzinski.com/2010/10/kanban-moving-cards-back.html

**Comment by flyingzumwalt** [comment]
I'll read these. Based on a cursory glance, 
* It seems that the purpose of WIP limits is to prevent/reduce multitasking. We should make sure this is a problem we need to solve before we apply a solution.
* This habit seems focused on situations where people check in with their teammates daily, not weekly. Do we need to modify the habit to fit our work cadence?
* What about contributors who have a day job and are working on open-source issues part-time?

**Comment by flyingzumwalt** [comment]
### Must Read

_If you're following this thread and haven't read all of the links @jbenet posted, read these ones:_

* The Best explanation of Kanban and its motivations from this list (it's really good): https://www.smartsheet.com/visualize-your-do-list-using-kanban-boards-optimize-workflow
* If you only read one of the links @jbenet posted, read that one because it's the most relevant to the current discussion: http://itsadeliverything.com/littles-law-the-basis-of-lean-and-kanban
* This post has a nice explanation of how Kanban allows you to signal team members' (and whole teams') availability to do work, and why it's designed that way: https://www.equinox.co.nz/blog/kanban-boards-done-columns-and-work-in-progress-limits

Here's how the first post explains WIP limits:

> The Doing column drives the Kanban Process. The only constraints in Kanban are Work-in-Progress (WIP) limits placed on the number of tasks that exist in this column at any one time. WIP limits, set by the team, should be realistic and not result in work overload for one or more team members. The number of tasks in the Doing column cannot exceed the established WIP limits; no new work can be entered into this column until the team completes the current work. This ensures that team members are always working on the highest-priority projects and do not deviate to other tasks.

### Are we formally Adopting Kanban? Should we?

The discussions about PM Process have explicitly said that we're not committing to a particular management methodology, but we keep adopting bits of Kanban process and now @jbenet is suggesting that we start limiting WIP. Limiting WIP is distinctively part of Kanban methodology. Is this a sign that we should consider formally adopting a Kanban approach rather than doing this piecemeal? It might be less disruptive and more rewarding if we just pull the trigger and try it properly.

### Is this Solving the Right Problem?

_Counterpoint to what I'm about to say: [this explanation of kanban](https://www.smartsheet.com/visualize-your-do-list-using-kanban-boards-optimize-workflow) frames WIP limits differently, and in a way that I do find compelling, so maybe I just need to learn more about Kanban._

I'm not convinced that we're suffering from the type of multitasking problem that most of these blog posts are solving.  Example from [this post](http://www.everydaykanban.com/2013/05/18/true-cost-of-no-wip-limits/):

> Do your stand-ups sound something like this?
> 
> Joe Developer: I started on that script, worked on it for a while and then I switched over to that web app i’ve been working on for the last few days. I got blocked on that so I started up that template they’re waiting on.
> 
> Jane Developer: I worked on that UI component library but then got pulled into a fire drill about something the PM needed right then. Today I’m going to finish the fire drill item, work on a defect someone asked me if I could sneak in really quick (its only 1hr of work) and then I’ll get back to the UI component.
> 
> … [repeat, repeat, repeat]…
> These same developers, [...] get really frustrated over the fact that “everything around here takes forever to get done.”

Our standups don't sound like that. As individuals we might grapple with the cost of multitasking & context switching (I often do) and as an org we have a lot of work to do in terms of building coherent milestones & roadmaps with strong narratives, but I don't see the team suffering from a lack of ability to focus on the tasks they've assigned to themselves.

If we're trying to reduce multitasking and context-switching, we should muffle Slack, Twitter and IRC. Limiting how long our GH issues are allowed to be open won't have much impact on context switching. Closing the nozzle on all the Slack/IRC/Facebook/Twitter chatter absolutely would reduce the context-switching.

### If implemented wrong, this would _increase_ context switching

@jbenet the links you posted are all about limiting the number of issues that each user (or team) can have "in progress" at any given time. That's a very different mechanism than the one you're proposing, which limits the amount of time that an issue can be "in progress". 

Putting this in the terms described in [this post about little's law](http://itsadeliverything.com/littles-law-the-basis-of-lean-and-kanban), 

> In other words, if you want to:
> 
> increase Throughput then limit WIP
> speed up the process, i.e. reduce Lead Time, then once again limit the WIP
> 
> Bit of a theme there.

would this proposed bot actually allow us to manipulate the WIP in that way?

Keep in mind today's discussion about Greenkeeper - a bot that's helpful in theory but annoying to the point of reduced productivity.

If a bot keeps bumping my issues out of the correct column, I have to either A) stop working and fix the label or B) ignore it and let the board be wrong. Option A forces me to switch to the mental context of "managing the board". Option B forces me to switch to the mental context of "figuring out which information on the board is actually accurate" whenever I want to pick up new tasks.

Alternatively - if we _were_ to start enforcing Kanban WIP limits in the style that all those Kanban blog posts propose, we would have to use completely new tools (or create even more bots). ie. Waffle doesn't support it, and most of us have work _in progress_ in many repositories at a time.  That said, there are probably tools out there for applying Kanban rules to Github projects.

**Comment by jbenet** [comment]
Quick notes:

> - This habit seems focused on situations where people check in with their teammates daily, not weekly. Do we need to modify the habit to fit our work cadence?
> - What about contributors who have a day job and are working on open-source issues part-time?

The full Kanban feel is tuned to an async-but-simultaneous team. I think with our boards, TZ spanning, looser teams, OSS values and habits, we would naturally yield a fully async (non simultaneous) cadence, even with several of the Kanban things.

> read these ones:

Thank you so much for picking out the best ones! 👍  i should endeavor to post links well-scoped, that way.

> The discussions about PM Process have explicitly said that we're not committing to a particular management methodology, but we keep adopting bits of Kanban process and now @jbenet is suggesting that we start limiting WIP. Limiting WIP is distinctively part of Kanban methodology. Is this a sign that we should consider formally adopting a Kanban approach rather than doing this piecemeal? It might be less disruptive and more rewarding if we just pull the trigger and try it properly.

Yeah, perhaps! I'm not sure what would be most helpful to others. It may be that -- as you say -- it is less disruptive and more rewarding to just do it properly. not sure.

> Our standups don't sound like that. As individuals we might grapple with the cost of multitasking & context switching (I often do) and as an org we have a lot of work to do in terms of building coherent milestones & roadmaps with strong narratives, but I don't see the team suffering from a lack of ability to focus on the tasks they've assigned to themselves.

I strongly disagree. I feel like our discussions sound somewhere between that, and a counterpart hyper productive group, but i suspect closer to that. Lots of people are overwhelmed by the number of things going on right now. There are signs to observe (people's shipping rate dropping, things getting stuck for a long time, discussions ambling without reaching resolutions, lowering of quality, things shipping with breakages, a loss of a sense of urgency, dependence and blockage (i'm blocking oh soooo many people! \o/. alarmingly many.), etc). A number of people have voiced this, too. Some people are doing a lot better now than they were a few weeks ago, and some people are doing worse. As a whole, it's something we're working on, but not good at, yet. Not with the scale of things we're doing. Using principles and tooling to help us focus, work on, and deliver individual tasks -- correctly and within reasonable time frame, and with a great ability to adapt and address issues as a team -- is something we need to get significantly better at. We're doing a lot better with the recent PMing improvments, and with the waffle boards, and I'm very excited to have a number of very important things shipping! \o/. But i'm not confident everyone will deliver on the agreed-upon milestones of our board. It did not happen last Qtr. I'm not sure how on-track people are right now.

---

> If we're trying to reduce multitasking and context-switching, we should muffle Slack, Twitter and IRC. 

Agreed on muffling slack, twitter. Maybe IRC-- lots of valuable points made in IRC. though i do wonder that people's super fast RTT probably means they can't focus much on their own stuff.

> Limiting how long our GH issues are allowed to be open won't have much impact on context switching.

Hm-- it's not about GH issues remaining open or closed-- they should remain open until they're done, as long as necessary. It's about limiting following explicit actions. The following is _an example_. I'm not proposing this as some unlikely panacea-- only that -- from my understanding of the `In Progress` column, and desires i've heard from several -- i think it can help us a lot.

Consider this:

```
0. Let X = be some important goal.
1. Start working on X.
2. I declare "I am currently working on X". (moving the card from "Ready" to "In Progress")
3. Maybe finish X.
4. Go to 0.
```
Pretty soon, people are working on a large number of things, with lots of loose threads consuming cognitive resources, not only on one person, but on several, and things aren't being completed ("flowing to done" in @haadcode's or Kanban's parlance). Instead, by adjusting to the following, we make it more clear -- to ourselves and to our team -- _when_ we're actually working on something, which allows us to focus better.

```
0. Let X = be some important goal.
1. Consider working on X. Is this the best thing to be working on? If so, _stop_ working on other things and refocus on X.
2. Start working on X.
3. I declare "I am currently working on X". (moving the card from "Ready" to "In Progress")
4. Maybe finish X.
5. If not, declare "I am not working on X anymore.". (moving the card from "In Progress" back to "Ready". If the card is still assigned, means: "I will pick it up later", or "Up for grabs" if not assigned.)
6. Go to 0.
```
While we do this _subconsciously_ to some degree, making it explicit allows us to rid ourselves of many multitasking, overcommitment, and miscommunication concerns. I don't _at all_ intend to replace that with a "busy work" thing that people dislike. What I intend is to ask people to make sure to have present what they're truly working on **right now**, what is **truly advancing**, what **is not advancing**, and that they allocate their conscious effort accordingly to the needed work. Background thought-based processes can happen separately, and i know many of us are very good at multitasking and doing many things semi-simultaneously. What I want is to give people the ability to be completely explicit about where their focus is, where their time is going, and where it's NOT going. In between states are pernicious and lead to much wasted time and effort.

---

> @jbenet the links you posted are all about limiting the number of issues that each user (or team) can have "in progress" at any given time. That's a very different mechanism than the one you're proposing, which limits the amount of time that an issue can be "in progress".

Yeah, great observation. It may be that limiting WIP period like this is counter productive. What I aimed for with this is:
- Remove stale `In Progress` labels (and establish a clear automated process for it)
- Externalize what people are currently doing internally (switching from one thing to another, putting something on hold while they work on something else (because of a blockage, or simply better fit))
- Then work towards clearly understanding and planning the amount that people are working on simultaneously. We can only do this with a good handle on establishing a real and accurate count, and a real and accurate sense of how much people can handle. (Not _directly_ arguing for limiting WIP as a direct rule, more arguing for tuning `count(WIP)` to match `count(capacity above thrashing)`. [Thrashing](https://en.wikipedia.org/wiki/Thrashing_(computer_science)) is bad.

It may be that this proposed bot is superfluous, or counter productive. Or that fitting what we're currently doing is not the right thing, and instead we should be focusing on limiting the WIP stack very explicitly. (im not sure what others think about it. the arguments in the posts seem very valid to me).

> If a bot keeps bumping my issues out of the correct column, I have to either A) stop working and fix the label or B) ignore it and let the board be wrong. Option A forces me to switch to the mental context of "managing the board". Option B forces me to switch to the mental context of "figuring out which information on the board is actually accurate" whenever I want to pick up new tasks.

Yeah great observations. I dont mean for busy work, and it may indeed turn out to be worse. This is part of why i'd like to bring this up. I do want to make sure whatever we do is _helpful_. This discussion -- in the first place -- clarified further that we're all not on the same page about what "In Progress" means, and that's already one issue that needs addressing. I think that aside from the flexibility afforded to other states, the `In Progress` and `Blocked` states -- if used -- **should** agree across projects.

What I don't want is people being simultaneously signed up to have 5-10 things all "In progress" and stalled from non-explicit abandonment, or stalled from cognitive complexity. [This is happening](https://github.com/search?p=1&q=org%3Aipfs+label%3A%22In+Progress%22&type=Issues), and [we're](https://github.com/ipfs/js-ipfs/milestones) [slipping](https://github.com/haadcode/orbit/milestones).

**Comment by kevina** [comment]
I have not been following this conversation that closely.  But if we re-purpose "in progress" to mean what someone is actively working on then I think we need some sort of other label for something that is is in really a long standing work in progress.  That is something that is not really ready yet.  If it where up to me I would leave "in progress" to mean something that is well in progress no mater what the time frame is and use a new label maybe "active" to mean something that is being worked on _now_.

[I also find the label "ready" to be very confusing.  But that is another discussion.]

Sorry if I am just adding noise to the conversation.

**Comment by flyingzumwalt** [comment]
@kevina regarding "ready", I want to work on making it clearer. In short, "ready" means "ready to be worked on". See [this line in the draft PM process doc](https://github.com/ipfs/pm/pull/183/files#diff-cca9cd45f93a569ea86d76223ca74498L182) and #188

**Comment by diasdavid** [comment]
Paging our DX team (@VictorBjelkholm & @travisperson) to have this thread in their radar.

**Comment by momack2** [comment]
This is still a great idea - but not something that is currently in scope or planned for foreseeable work in the project operations group, so I'm going to close it and we can resurrect if we see a large need in any of our groups.
